src/main/kotlin/com/openlattice/mechanic/upgrades/SyncOrgEntitySetsMissingInMeta.kt

```diff
@@ -0,0 +1,128 @@
+import com.openlattice.organizations.Organization
+import com.openlattice.organizations.OrganizationMetadataEntitySetsService
+import org.apache.olingo.commons.api.edm.FullQualifiedName
+import org.slf4j.LoggerFactory
+import java.util.*
+
+class SyncOrgEntitySetsMissingInMeta(
+    private val toolbox: Toolbox,
+    private val metadataEntitySetsService: OrganizationMetadataEntitySetsService,
+    private val dataService: DataGraphManager
+) : Upgrade {
+
+    companion object {
+        private val logger = LoggerFactory.getLogger(SyncOrgEntitySetsMissingInMeta::class.java)
+
+        private val ID_FQN = FullQualifiedName("ol.id")
+        private val OL_ID_UUID = UUID.fromString("39e13db7-a730-421a-a600-ae0674060140")
+    }
+
+    override fun getSupportedVersion(): Long {
+        return Version.V2021_03_10.value
+    }
+
+    override fun upgrade(): Boolean {
+
+        val orgs = HazelcastMap.ORGANIZATIONS.getMap(toolbox.hazelcast).toMap()
+        val entitySetsByOrg = toolbox.entitySets.values.groupBy { it.organizationId }
+
+        val idPT = toolbox.propertyTypes[OL_ID_UUID]
+        if (idPT == null) {
+            logger.error("PropertyType \"ol.id\" is null")
+            return false
+        }
+
+        orgs.values.forEachIndexed { index, org ->
+
+            logger.info("================================")
+            logger.info("================================")
+            logger.info("starting processing org ${org.id}")
+
+            val orgEntitySetsIds = entitySetsByOrg.getOrDefault(org.id, listOf()).mapNotNull { it.id }.toSet()
+            if (orgEntitySetsIds.isEmpty()) {
+                logger.warn("org does not have entity sets")
+                return@forEachIndexed
+            }
+
+            val dataSetsESID = org.organizationMetadataEntitySetIds.datasets
+            val dsEntitySet = toolbox.entitySets[dataSetsESID]
+            if (dsEntitySet == null) {
+                logger.warn("\"datasets\" meta entity set is null")
+                return@forEachIndexed
+            }
+
+            val data = dataService.getEntitySetData(
+                mapOf(dataSetsESID to Optional.empty()),
+                linkedSetOf(idPT.type.toString()),
+                mapOf(dataSetsESID to mapOf(idPT.id to idPT)),
+                dsEntitySet.isLinking
+            )
+
+            if (data.entities == null) {
+                logger.info("getEntitySetData returned null entities")
+                return@forEachIndexed
+            }
+
+            val ids = data.entities
+                .filterNotNull()
+                .mapNotNull { entity ->
+                    val id = entity[ID_FQN]?.first()
+                    try {
+                        UUID.fromString(id.toString())
+                    } catch (e: Exception) {
+                        logger.error("invalid data set id $id", e)
+                        return@mapNotNull null
+                    }
+                }
+                .toSet()
```

Comment:
note: Can also do:
```
.mapNotNullTo( mutableSetOf<>() ) {
    // mapping code 
}
```

instead of 

```
.mapNotNull { 
  // mapping code 
}.toSet()
```

Reply:
so fancy!